neurokernel/mpi_run.py

```diff
@@ -0,0 +1,227 @@
+    log_file_name : str
+        File to send log messages to.
+    
+    Returns
+    -------
+    output : str
+        The stdout from the function run with mpiexec cast to a string.
+
+    Usage
+    -----
+    Does not seem to work with openmpi version 2
+    func should not import neurokernel.mpi_relaunch
+    All modules and variables used must be imported or defined within func
+    Returns the stdout from the function run under 'mpiexec -np 1 python {tmp_file_name}'
+    """
+
+    l = LoggerMixin("mpi_run()",log_on=log)
+    
+    if callable(func):
+        func_text = inspect.getsource(func)
+        # Make a feeble attempt at fixing indentation. Will work for a nested function
+        # that takes no args, not a member function that expects (self) or a class
+        func_text = "\n" + re.sub(r"(^\s+)def ","def ",func_text) + "\n"
+        func_name = func.__name__
+    else: 
+        func_text = "\n" + func + "\n"
+        func_name = re.search('def *(.*)\(\):', func_text).group(1)
+
+    target_text  = "\n"
+    
+    if targets:
+        for t in targets:
+            target_text +=  "\n" + inspect.getsource(t) + "\n"
+
+    main_code  = "\n"
+    main_code += "\nif __name__ == \"__main__\":"
+    main_code += "\n   import neurokernel.mpi as mpi"
+    main_code += "\n   from neurokernel.mixins import LoggerMixin"
+    main_code += "\n   from mpi4py import MPI"
+
+    if log:
+        main_code += "\n   mpi.setup_logger(screen=%s, file_name=\"%s\"," % (log_screen, log_file_name)
+        main_code += "\n                    mpi_comm=MPI.COMM_WORLD, multiline=True)"
+
+    main_code += "\n   l = LoggerMixin(\"%s\",%s)" % (func_name,str(log))
+    main_code += "\n   try:"
+    main_code += "\n      %s()" % func_name
+    main_code += "\n      print(\"MPI_RUN_SUCCESS: %s\")" % func_name
+    main_code += "\n      l.log_info(\"MPI_RUN_SUCCESS: %s\")" % func_name
+    main_code += "\n   except Exception as e:"
+    main_code += "\n      print(\"MPI_RUN_FAILURE: %s\")" % func_name
+    main_code += "\n      l.log_error(\"MPI_RUN_FAILURE: %s\")" % func_name
+    main_code += "\n      print(e)"
+    main_code += "\n"
+
+    try:
+        from mpi4py import MPI
+        #Write code for the function to a temp file
+        temp = tempfile.NamedTemporaryFile(delete = delete_tempfile)
+        temp.write(target_text)
+        temp.write(func_text)
+        temp.write(main_code)
+        temp.flush()
+        
+        #Execute the code
+        #There's a bug in Open MPI v2 that prevents running this with mpiexec. Running 'from mpi4py import MPI' 
+        #does a basic mpi_relaunch which will work for the notebook code, but you give up some of the features 
+        #of mpiexec.
+        if MPI.Get_library_version().startswith("Open MPI v2"):
+            command = ["python", temp.name]
+        else:
+            command = ["mpiexec", "-np", "1", "python", temp.name]
+
+        env = os.environ.copy()
```

Comment:
Requires to remove some SLURM related env entries to prevent it from preventing mpiexec from starting multiple processes (see mpi_relaunch.py):
```
for k in env.keys():
    if k.startswith('SLURM'):
        del env[k]
```